kvcoord: fix the write buffer's adjustError

miraradeva opened the pull request:
Previously, `adjustError` handled the cases of stripped requests (not sent to KV at all) and non-transformed requests (sent directly to KV), but didn't always handle the cases of transformed requests correctly.

This commit simplifies the logic in `adjustError` and adds additional testing.

Fixes: [#144694](https://github.com/cockroachdb/cockroach/issues/144694)

Release note: None

miraradeva committed:
kvcoord: fix the write buffer's adjustError

Previously, `adjustError` handled the cases of stripped requests
(not sent to KV at all) and non-transformed requests (sent directly to
KV), but didn't always handle the cases of transformed requests
correctly.

This commit simplifies the logic in `adjustError` and adds additional
testing.

Fixes: #144694

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146120)
<!-- Reviewable:end -->

miraradeva commented on pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go:
```diff
@@ -452,39 +452,35 @@ func (twb *txnWriteBuffer) adjustError(
+				// [1] Some requests are always collected as transformations, but
+				// they're never transformed. Attributing an error to them shouldn't
+				// confuse the client.
+				transformed := curTs.origRequest.Method() == kvpb.Put ||
```
Listing these request types here seems a little fragile, but we're probably not planning on adding more transformed types?

Alternatively, we can add a field in the `transformation` struct to denote if a request was actually transformed.

stevendanna reviewed (commented):
Nice! Thanks for finding & fixing this bug along the way.  

I wonder if KVNemesis would have caught this?  I think not because do don't do much validation of what a scan should have returned.  Perhaps an interesting test for the future (not this PR) would be something that generates random txn's at the KV level and asserts that we get the same results with and without buffered writes enabled. I think it is hard to hit this via SQL because it is hard (maybe impossible) to create a batch that has both scans and writes in it.

miraradeva commented on pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go:
```diff
@@ -675,13 +675,25 @@ func (twb *txnWriteBuffer) applyTransformations(
 	var ts transformations
```
I'll add a TODO about the allocations. Next on my list is to add a benchmark and then tackle some of these allocation improvements.

I agree renaming the struct would be nice. Especially after I add a new field to track whether the request was transformed or not. I'll do it in this PR, in a separate commit.

miraradeva commented:
> I wonder if KVNemesis would have caught this? I think not because do don't do much validation of what a scan should have returned. Perhaps an interesting test for the future (not this PR) would be something that generates random txn's at the KV level and asserts that we get the same results with and without buffered writes enabled. I think it is hard to hit this via SQL because it is hard (maybe impossible) to create a batch that has both scans and writes in it.

I wondered the same thing. The validator is technically supposed to validate read your own writes. This test below is for Get but we could do the same for Scan I think:

https://github.com/cockroachdb/cockroach/blob/c83c57d354741ac36740894f7387c014eb6c09fe/pkg/kv/kvnemesis/validator_test.go#L1050-L1060

Are buffered writes always on for kvnemesis?

miraradeva committed:
kvcoord: simplify mergeResponseWithRequestRecords

Now that all requests are tracked in the write buffer, regardless of
whether they were transformed, we can simplify the iteration in
`mergeResponseWithRequestRecords`, similarly to `adjustError`, and just
iterate over all request records.

Part of: #146103

Release note: None

yuzefovich reviewed (approved):
Nice cleanup! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> I think it is hard to hit this via SQL because it is hard (maybe impossible) to create a batch that has both scans and writes in it.

The only scenario that came to mind is the "INSERT fast-path" where we will include Puts in the same batch as reads for FK and UNIQUE checks, but those reads will be of different keys than we had in Puts, so I don't think we'll ever have a batch from SQL that has writes and reads of the same keys.

> Reviewed 1 of 2 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146120#-:-OPg_kWPBzDCL44w1Ua0:b-qzj82q)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @miraradeva, and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 457 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/146120#-OPgbIkUELB9kNKTeEZw:-OPgbIkUELB9kNKTeEZx:b-gcgy24) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d45296250db74e6bbe3e935df906af83d78c8b8/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L457)):*
> ```Go
> 		baIdx := int32(0)
> 		// Note that all requests in the batch are guaranteed to be in the list of
> 		// transformations.
> ```

nit: `s/transformations/requestRecords/`?
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 412 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/146120#-OPgakAw8rRL3dqJ5XGb:-OPgakAw8rRL3dqJ5XGc:b-cxzako) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d45296250db74e6bbe3e935df906af83d78c8b8/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L412)):*
> ```Go
> // of the batch is buffered on the client.
> //
> // TODO(arul): extend this test to transformations as well once we start
> ```

nit: this TODO seems now stale.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146120)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @miraradeva)



<!-- Sent from Reviewable.io -->

miraradeva commented:
bors r=yuzefovich,stevendanna
